### PR TITLE
Making HAL command buffers take buffers as indirect args.

### PR DESCRIPTION
**Pull request by benvanik**
This allows for the same command buffer interface to be used for recording indirect command buffers that reference buffer table slots instead of having concrete iree_hal_buffer_t pointers available at the time they are recorded. As part of this first change behavior is kept largely the same and only the arguments are changed. Future changes per backend will be needed to actually support driver-side encoding of the indirect bindings.

### CODE REVIEW

File: runtime/src/iree/hal/command_buffer.h [hunk 1]
```diff
@@ -725,8 +685,7 @@ IREE_API_EXPORT iree_status_t iree_hal_command_buffer_push_constants(
 IREE_API_EXPORT iree_status_t iree_hal_command_buffer_push_descriptor_set(
     iree_hal_command_buffer_t* command_buffer,
     iree_hal_pipeline_layout_t* pipeline_layout, uint32_t set,
-    iree_host_size_t binding_count,
-    const iree_hal_descriptor_set_binding_t* bindings);
+    iree_host_size_t binding_count, const iree_hal_buffer_ref_t* bindings);
```
Comment (ScottTodd) [inline runtime/src/iree/hal/command_buffer.h:696]: I found these diffs in the CTS test usage confusing, since `iree_hal_command_buffer_push_descriptor_set()` no longer takes an argument with `descriptor_set` in the name:
```diff
-  iree_hal_descriptor_set_binding_t descriptor_set_bindings[] = {
+  iree_hal_buffer_ref_t descriptor_set_bindings[] = {
```

Looks like the `iree_hal_descriptor_set_binding_t` struct got merged into `iree_hal_buffer_ref_t`.

How about updating the docstring comment for this function to more directly associate "buffer bindings" with "descriptor sets"? Could also rename the parameter and/or struct.

File: tools/iree-benchmark-executable-main.c
Comment (ScottTodd) [inline tools/iree-benchmark-executable-main.c:1]: ASan didn't like something in this?

https://github.com/iree-org/iree/actions/runs/9845575632/job/27181406817?pr=17730#step:4:10181
```
 FAIL: IREE :: test/iree-benchmark-executable.mlir (1 of 1)
  ******************** TEST 'IREE :: test/iree-benchmark-executable.mlir' FAILED ********************
  Exit Code: 1
  
  Command Output (stderr):
  --
  RUN: at line 17: iree-compile      --compile-mode=hal-executable      --iree-hal-target-backends=vmvx      /work/tools/test/iree-benchmark-executable.mlir |  iree-benchmark-executable      --device=local-sync      --executable_format=vmvx-bytecode-fb      --executable_file=-      --entry_point=0      --binding=512xf32      --binding=512xf32      --binding=512xf32      --workgroup_count=1,1,1      --workgroup_count=512,1,1 |  FileCheck /work/tools/test/iree-benchmark-executable.mlir
  + iree-compile --compile-mode=hal-executable --iree-hal-target-backends=vmvx /work/tools/test/iree-benchmark-executable.mlir
  + iree-benchmark-executable --device=local-sync --executable_format=vmvx-bytecode-fb --executable_file=- --entry_point=0 --binding=512xf32 --binding=512xf32 --binding=512xf32 --workgroup_count=1,1,1 --workgroup_count=512,1,1
  + FileCheck /work/tools/test/iree-benchmark-executable.mlir
  2024-07-08T19:27:39+00:00
  Running iree-benchmark-executable
  Run on (96 X 2000.17 MHz CPU s)
  CPU Caches:
    L1 Data 32 KiB (x48)
    L1 Instruction 32 KiB (x48)
    L2 Unified 1024 KiB (x48)
    L3 Unified 39424 KiB (x2)
  Load Average: 79.96, 82.15, 39.39
  ***WARNING*** Library was built as DEBUG. Timings may be affected.
  
  =================================================================
  ==60881==ERROR: LeakSanitizer: detected memory leaks
  
  Direct leak of 4896 byte(s) in 2 object(s) allocated from:
      #0 0x55e3af127509 in calloc (/work/build-asan/tools/iree-benchmark-executable+0x137509) (BuildId: 939f2912d8f1c87f1adc5b03fb006a7d69a7e99b)
      #1 0x55e3af16c440 in iree_allocator_system_alloc /work/runtime/src/iree/base/allocator.c:105:17
      #2 0x55e3af16c440 in iree_allocator_system_ctl /work/runtime/src/iree/base/allocator.c:145:14
      #3 0x55e3af16be09 in iree_allocator_issue_alloc /work/runtime/src/iree/base/allocator.c:27:10
      #4 0x55e3af16be09 in iree_allocator_malloc /work/runtime/src/iree/base/allocator.c:32:10
      #5 0x55e3af34b920 in iree_hal_inline_command_buffer_create /work/runtime/src/iree/hal/local/inline_command_buffer.c:163:26
      #6 0x55e3af277f25 in iree_hal_sync_device_create_command_buffer /work/runtime/src/iree/hal/drivers/local_sync/sync_device.c:237:12
      #7 0x55e3af16b7fd in iree_benchmark_executable_run /work/tools/iree-benchmark-executable-main.c:261:5
      #8 0x55e3af19648f in iree_benchmark_run(char const*, iree_benchmark_def_t const*, benchmark::State&) /work/runtime/src/iree/testing/benchmark_full.cc:114:26
      #9 0x55e3af19648f in iree_benchmark_register::$_0::operator()(benchmark::State&) const /work/runtime/src/iree/testing/benchmark_full.cc:134:9
      #10 0x55e3af19648f in benchmark::internal::LambdaBenchmark<iree_benchmark_register::$_0>::Run(benchmark::State&) /work/third_party/benchmark/include/benchmark/benchmark.h:1380:44
      #11 0x55e3af2313bb in benchmark::internal::BenchmarkInstance::Run(long, int, benchmark::internal::ThreadTimer*, benchmark::internal::ThreadManager*, benchmark::internal::PerfCountersMeasurement*) const /work/third_party/benchmark/src/benchmark_api_internal.cc:98:14
      #12 0x55e3af2028c3 in benchmark::internal::(anonymous namespace)::RunInThread(benchmark::internal::BenchmarkInstance const*, long, int, benchmark::internal::ThreadManager*, benchmark::internal::PerfCountersMeasurement*) /work/third_party/benchmark/src/benchmark_runner.cc:132:10
      #13 0x55e3af201946 in benchmark::internal::BenchmarkRunner::DoNIterations() /work/third_party/benchmark/src/benchmark_runner.cc:273:3
      #14 0x55e3af203e81 in benchmark::internal::BenchmarkRunner::DoOneRepetition() /work/third_party/benchmark/src/benchmark_runner.cc:424:9
      #15 0x55e3af1a170d in benchmark::internal::(anonymous namespace)::RunBenchmarks(std::vector<benchmark::internal::BenchmarkInstance, std::allocator<benchmark::internal::BenchmarkInstance>> const&, benchmark::BenchmarkReporter*, benchmark::BenchmarkReporter*) /work/third_party/benchmark/src/benchmark.cc:438:14
      #16 0x55e3af1a170d in benchmark::RunSpecifiedBenchmarks(benchmark::BenchmarkReporter*, benchmark::BenchmarkReporter*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>) /work/third_party/benchmark/src/benchmark.cc:611:5
      #17 0x55e3af1a0281 in benchmark::RunSpecifiedBenchmarks() /work/third_party/benchmark/src/benchmark.cc:535:10
      #18 0x55e3af16b24c in iree_benchmark_executable_from_flags /work/tools/iree-benchmark-executable-main.c:491:3
      #19 0x55e3af16b24c in main /work/tools/iree-benchmark-executable-main.c:563:26
      #20 0x7fac31e5dd8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 490fef8403240c91833978d494d39e537409b92e)
  
  Direct leak of 176 byte(s) in 2 object(s) allocated from:
      #0 0x55e3af127509 in calloc (/work/build-asan/tools/iree-benchmark-executable+0x137509) (BuildId: 939f2912d8f1c87f1adc5b03fb006a7d69a7e99b)
      #1 0x55e3af16c440 in iree_allocator_system_alloc /work/runtime/src/iree/base/allocator.c:105:17
      #2 0x55e3af16c440 in iree_allocator_system_ctl /work/runtime/src/iree/base/allocator.c:145:14
      #3 0x55e3af16be09 in iree_allocator_issue_alloc /work/runtime/src/iree/base/allocator.c:27:10
      #4 0x55e3af16be09 in iree_allocator_malloc /work/runtime/src/iree/base/allocator.c:32:10
      #5 0x55e3af27a688 in iree_hal_sync_semaphore_create /work/runtime/src/iree/hal/drivers/local_sync/sync_semaphore.c:74:26
      #6 0x55e3af16b680 in iree_benchmark_executable_run /work/tools/iree-benchmark-executable-main.c:224:3
      #7 0x55e3af19648f in iree_benchmark_run(char const*, iree_benchmark_def_t const*, benchmark::State&) /work/runtime/src/iree/testing/benchmark_full.cc:114:26
      #8 0x55e3af19648f in iree_benchmark_register::$_0::operator()(benchmark::State&) const /work/runtime/src/iree/testing/benchmark_full.cc:134:9
      #9 0x55e3af19648f in benchmark::internal::LambdaBenchmark<iree_benchmark_register::$_0>::Run(benchmark::State&) /work/third_party/benchmark/include/benchmark/benchmark.h:1380:44
      #10 0x55e3af2313bb in benchmark::internal::BenchmarkInstance::Run(long, int, benchmark::internal::ThreadTimer*, benchmark::internal::ThreadManager*, benchmark::internal::PerfCountersMeasurement*) const /work/third_party/benchmark/src/benchmark_api_internal.cc:98:14
      #11 0x55e3af2028c3 in benchmark::internal::(anonymous namespace)::RunInThread(benchmark::internal::BenchmarkInstance const*, long, int, benchmark::internal::ThreadManager*, benchmark::internal::PerfCountersMeasurement*) /work/third_party/benchmark/src/benchmark_runner.cc:132:10
      #12 0x55e3af201946 in benchmark::internal::BenchmarkRunner::DoNIterations() /work/third_party/benchmark/src/benchmark_runner.cc:273:3
      #13 0x55e3af203e81 in benchmark::internal::BenchmarkRunner::DoOneRepetition() /work/third_party/benchmark/src/benchmark_runner.cc:424:9
      #14 0x55e3af1a170d in benchmark::internal::(anonymous namespace)::RunBenchmarks(std::vector<benchmark::internal::BenchmarkInstance, std::allocator<benchmark::internal::BenchmarkInstance>> const&, benchmark::BenchmarkReporter*, benchmark::BenchmarkReporter*) /work/third_party/benchmark/src/benchmark.cc:438:14
      #15 0x55e3af1a170d in benchmark::RunSpecifiedBenchmarks(benchmark::BenchmarkReporter*, benchmark::BenchmarkReporter*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char>>) /work/third_party/benchmark/src/benchmark.cc:611:5
      #16 0x55e3af1a0281 in benchmark::RunSpecifiedBenchmarks() /work/third_party/benchmark/src/benchmark.cc:535:10
      #17 0x55e3af16b24c in iree_benchmark_executable_from_flags /work/tools/iree-benchmark-executable-main.c:491:3
      #18 0x55e3af16b24c in main /work/tools/iree-benchmark-executable-main.c:563:26
      #19 0x7fac31e5dd8f  (/lib/x86_64-linux-gnu/libc.so.6+0x29d8f) (BuildId: 490fef8403240c91833978d494d39e537409b92e)
  
  SUMMARY: AddressSanitizer: 5072 byte(s) leaked in 4 allocation(s).
  
  --
  
  ********************
  ********************
  Failed Tests (1):
    IREE :: test/iree-benchmark-executable.mlir
```

**Review by ScottTodd — COMMENTED** [review]
"Command buffers taking buffers as indirect args" is different from "indirect command buffers", right? https://iree.dev/developers/design-docs/design-roadmap/#indirect-command-bufferon-accelerator-execution (was checking if we could mark some items there / elsewhere as in-progress or finished with this body of work)

File: tools/iree-benchmark-executable-main.c
Comment (benvanik) [reply tools/iree-benchmark-executable-main.c:1]: benchmark-executable was failing due to a bug and it doesn't cleanly exit on failures

**Comment by benvanik** [comment]
Good q; it's a generic description of a primitive: an indirect command buffer is any command buffer that indirectly reference resources/configuration. Device-side dispatch is also a generic description and of something else (dispatching logic from the device itself without host involvement). The two can be used together e.g. host records an indirect command buffer and then device submits it with resources determined on-device during execution, etc. But device-side dispatch need not use command buffers at all and indirect command buffers can be used for host-only submission schemes. See https://developer.apple.com/documentation/metal/indirect_command_encoding as it has some useful breakdowns (ours isn't exactly the same as that's focused on rendering and such, but same ideas).

In other words: this is unrelated to device-side dispatch but may be used by it when implemented.

File: runtime/src/iree/hal/command_buffer.h (hunk 1 above)
Comment (benvanik) [reply runtime/src/iree/hal/command_buffer.h:696]: did I just get permission to increase scope? I think I did! ;P
I may do a follow-up that renames `push_descriptor_set` to `push_bindings`. Descriptor set layouts are still used by pipeline layouts but we don't have a HAL concept of a descriptor set resource so it doesn't make much sense beyond matching Vulkan naming.

For now I'll keep it the same but added a description to the comment.

Comment (ScottTodd) [reply runtime/src/iree/hal/command_buffer.h:696]: Scope creep yay!

SGTM

File: runtime/src/iree/hal/command_buffer.h [hunk 2]
```diff
@@ -715,18 +675,25 @@ IREE_API_EXPORT iree_status_t iree_hal_command_buffer_push_constants(
     iree_hal_pipeline_layout_t* pipeline_layout, iree_host_size_t offset,
     const void* values, iree_host_size_t values_length);
 
-// Pushes a descriptor set and associates it with |set|.
+// Pushes descriptor set bindings and associates it with |set|.
```
Comment (ScottTodd) [inline runtime/src/iree/hal/command_buffer.h]: grammar nit

```suggestion
// Pushes descriptor set bindings and associates them with |set|.
```

**Review by ScottTodd — APPROVED** [review]

**Comment by maxbartel** [comment]
@benvanik I think this broke the metal build... 
```
runtime/src/iree/hal/drivers/metal/CMakeFiles/iree_hal_drivers_metal_metal.objects.dir/direct_command_buffer.m.o 
/usr/bin/clang  -I/Users/bartel/Documents/iree -I/Users/bartel/Documents/iree/iree-build -I/Users/bartel/Documents/iree/runtime/src -I/Users/bartel/Documents/iree/iree-build/runtime/src -I/Users/bartel/Documents/iree/iree-build/runtime/src/iree/base/internal/flatcc -I/Users/bartel/Documents/iree/iree-build/runtime/src/iree/schemas -isystem /Users/bartel/Documents/iree/third_party/flatcc/include -O2 -g   -gsplit-dwarf -ggnu-pubnames -std=gnu11 -arch arm64 -fPIC -fvisibility=hidden -Werror -Wall -Wno-error=deprecated-declarations -Wno-ambiguous-member-template -Wno-char-subscripts -Wno-extern-c-compat -Wno-gnu-alignof-expression -Wno-gnu-variable-sized-type-not-at-end -Wno-ignored-optimization-argument -Wno-invalid-offsetof -Wno-invalid-source-encoding -Wno-mismatched-tags -Wno-pointer-sign -Wno-reserved-user-defined-literal -Wno-return-type-c-linkage -Wno-self-assign-overloaded -Wno-sign-compare -Wno-signed-unsigned-wchar -Wno-strict-overflow -Wno-trigraphs -Wno-unknown-pragmas -Wno-unknown-warning-option -Wno-unused-command-line-argument -Wno-unused-const-variable -Wno-unused-function -Wno-unused-local-typedef -Wno-unused-private-field -Wno-user-defined-warnings -Wno-missing-braces -Wctad-maybe-unsupported -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wformat-security -Wgnu-redeclared-enum -Wimplicit-fallthrough -Winfinite-recursion -Wliteral-conversion -Wnon-virtual-dtor -Woverloaded-virtual -Wpointer-arith -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety -Wthread-safety-beta -Wunused-comparison -Wvla -fno-lax-vector-conversions -fno-omit-frame-pointer -fmacro-prefix-map=/Users/bartel/Documents/iree=iree -I/Users/bartel/Documents/iree/third_party/flatcc/include/ -I/Users/bartel/Documents/iree/third_party/flatcc/include/flatcc/reflection/ -MD -MT runtime/src/iree/hal/drivers/metal/CMakeFiles/iree_hal_drivers_metal_metal.objects.dir/direct_command_buffer.m.o -MF runtime/src/iree/hal/drivers/metal/CMakeFiles/iree_hal_drivers_metal_metal.objects.dir/direct_command_buffer.m.o.d -o runtime/src/iree/hal/drivers/metal/CMakeFiles/iree_hal_drivers_metal_metal.objects.dir/direct_command_buffer.m.o -c /Users/bartel/Documents/iree/runtime/src/iree/hal/drivers/metal/direct_command_buffer.m
/Users/bartel/Documents/iree/runtime/src/iree/hal/drivers/metal/direct_command_buffer.m:543:72: error: no member named 'buffer' in 'struct iree_hal_buffer_barrier_t'
          iree_hal_buffer_allocated_buffer(segment->buffer_barriers[i].buffer));
                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
/Users/bartel/Documents/iree/runtime/src/iree/hal/drivers/metal/direct_command_buffer.m:671:70: error: use of undeclared identifier 'target_buffer'; did you mean 'target_offset'?
      iree_hal_resource_set_insert(command_buffer->resource_set, 1, &target_buffer);
                                                                     ^~~~~~~~~~~~~
                                                                     target_offset
/Users/bartel/Documents/iree/runtime/src/iree/hal/drivers/metal/direct_command_buffer.m:645:22: note: 'target_offset' declared here
  iree_device_size_t target_offset =
                     ^
/Users/bartel/Documents/iree/runtime/src/iree/hal/drivers/metal/direct_command_buffer.m:1183:26: error: incompatible function pointer types initializing 'iree_status_t (*)(iree_hal_command_buffer_t *, iree_hal_executable_t *, int32_t, iree_hal_buffer_ref_t)' (aka 'struct iree_status_handle_t *(*)(struct iree_hal_command_buffer_t *, struct iree_hal_executable_t *, int, struct iree_hal_buffer_ref_t)') with an expression of type 'iree_status_t (iree_hal_command_buffer_t *, iree_hal_executable_t *, int32_t, iree_hal_buffer_ref_t, iree_device_size_t)' (aka 'struct iree_status_handle_t *(struct iree_hal_command_buffer_t *, struct iree_hal_executable_t *, int, struct iree_hal_buffer_ref_t, unsigned long)') [-Wincompatible-function-pointer-types]
    .dispatch_indirect = iree_hal_metal_command_buffer_prepare_dispatch_indirect,
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
3 errors generated.
```

**Comment by benvanik** [comment]
Looks like it - hooray for having no CI bots for mac at all - mac support is effectively best-effort until then :(
I'll break out my mac and get it fixed in a bit.

**Comment by benvanik** [comment]
0c90e5ee91c62999a03c11c9a0b978920304bc2d should fix the build break. Thanks for the heads-up!

**Comment by ScottTodd** [comment]
Adding a CI build that covers the Metal HAL in https://github.com/iree-org/iree/pull/17835

**Comment by maxbartel** [comment]
Yes, this fixes the build. Thank you @benvanik! 🙂
